### ISSUE TITLE
DROP TABLE fails with "Failed to translate view" if executed against a broken view

### ISSUE THREAD

**Issue by hashhar** [trinodb/trino#9023]

**Comment by hashhar** [comment]
Note that this problem is not limited to views. This is a more general problem where checking for existence of other objects (tables, views, materialized views) can fail even though the original statement wasn't going to fail.

IMO if the expected user-facing effect of https://github.com/trinodb/trino/commit/7dff33590ec7d3f0584d608a67054e9e7100502c#diff-0b163286d3b9bbd8a73ae6c404e8e2487077b0266b6ffb8a34d5b8a8d2c045afR70 was to provide better error messages then we should ignore VIEW_TRANSLATION_ERROR and TABLE_NOT_FOUND - same as we do in `HiveMetadata#getViews` (see https://github.com/trinodb/trino/blob/3b7a3425c286e0461fb41268614bfb458f6fbe43/plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java#L1968-L1977)

**Comment by findepi** [comment]
cc @lukasz-stec @sopel39

**Comment by sopel39** [comment]
Why would you use `DROP TABLE IF EXISTS two;` on `two`, which is a view? 
You should use `DROP VIEW two` instead. Does that succeed?

**Comment by lukasz-stec** [comment]
I see the getViiew can also fail with HiveViewNotSupportedException
```
if (!translateHiveViews && !isPrestoView(view)) {
                        throw new HiveViewNotSupportedException(viewName);
                    }
```
we should probably ignore that as well.

One option I see is to add 
```
boolean viewExists(ConnectorSession session, SchemaTableName viewName);
boolean materializedViewExists(ConnectorSession session, SchemaTableName viewName);
```
that would, for some connectors (hive, iceberg,) ignore some known exceptions.
I think it's fine if we fail for unexpected reasons (e.g. network is down)

**Comment by sopel39** [comment]
> IMO if the expected user-facing effect of 7dff335#diff-0b163286d3b9bbd8a73ae6c404e8e2487077b0266b6ffb8a34d5b8a8d2c045afR70 was to provide better error messages then we should ignore VIEW_TRANSLATION_ERROR 

`VIEW_TRANSLATION_ERROR` is not part of SPI, should it be?

> Caused by: org.apache.calcite.runtime.CalciteContextException: At line 0, column 0: Object 'one' not found within 'hive.default'

Ideally, it's the engine that performs analysis/translation and analysis is not being done normally when view handle is being obtained, so it doesn't fail during view removal.

I think if we want to fix it, then there are two approaches (maybe more?):
1. We expose `VIEW_TRANSLATION_ERROR` to SPI and catch it when view handle is being obtained.
2. We extract SQL translation out of connector back to SPI somehow (e.g `Optional<String> Metadata#translateViewQuery(String nativeQuery)`)

IMO I feel that our in-connector translation is fragile and somewhat shortcut, because now connector has to be aware of engine SQL syntax and features. Yet I'm not sure what the better approach should be.

cc @martint

**Comment by sopel39** [comment]
In any case, `DROP TABLE` on a view should fail regardless, because it is a different entity even if `DROP TABLE` works accidentally.

**Comment by hashhar** [comment]
> In any case, `DROP TABLE` on a view should fail regardless, because it is a different entity even if `DROP TABLE` works accidentally.

Yes, this is only about the error message being misleading. I think the new behaviour is correct.

**Comment by findepi** [comment]
> In any case, `DROP TABLE` on a view should fail regardless, because it is a different entity even if `DROP TABLE` works accidentally.

Does the `DROP VIEW` succeed on a view that cannot be translated?

**Comment by sopel39** [comment]
`DROP TABLE view IF EXSITS` should also succeed, but it will fail now for Hive if view cannot be translated.

**Comment by findepi** [comment]
> `DROP TABLE view IF EXSITS` should also succeed, but it will fail now for Hive if view cannot be translated.

I feel like `DROP TABLE IF EXISTS a_view` should **not** succeed. 
And we don't have `DROP TABLE IF IT IS TABLE ...` statement.

What does the spec say? cc @martint

**Comment by findepi** [comment]
> I feel like `DROP TABLE IF EXISTS a_view` should **not** succeed.

Continued under separate issue https://github.com/trinodb/trino/issues/9129

**Comment by findepi** [comment]
> Does the `DROP VIEW` succeed on a view that cannot be translated?

It fails. I believe this is the biggest problem here (since DROP TABLE should fail _anyway_, it's of secondary nature how nicely it fails). 
I updated the issue description accordingly.


The problem seems to be that the only way for engine to determine the view existence is by `io.trino.spi.connector.ConnectorMetadata#getView`, and this includes the view definition. In Hive complex views' case, getting view definition is not a trivial operation and may fail.
@losipiuk should such use-case be considered as related to https://github.com/trinodb/trino/issues/6551, or do you think we should solve it separately?

**Comment by losipiuk** [comment]
> > Does the `DROP VIEW` succeed on a view that cannot be translated?
> 
> It fails. I believe this is the biggest problem here (since DROP TABLE should fail _anyway_, it's of secondary nature how nicely it fails).
> I updated the issue description accordingly.
> 
> The problem seems to be that the only way for engine to determine the view existence is by `io.trino.spi.connector.ConnectorMetadata#getView`, and this includes the view definition. In Hive complex views' case, getting view definition is not a trivial operation and may fail.
> @losipiuk should such use-case be considered as related to #6551, or do you think we should solve it separately?

With new APIs which eventually will (hopefully) come out of #6551 this could be resolved. So I see it as related. Though, given that there is no ETA on #6551, and the fact we are unable to DROP malformed views at the moment, I think this one should be addressed separately for now.
